### PR TITLE
Upgrade Badger  v3->Badger  v4

### LINKED ISSUE #5611: Upgrade Badger storage to v4
Renovate bot created this PR https://github.com/jaegertracing/jaeger/pull/5610, but it's not really doing anything since we'd also need to change the imports:
```
$ grx badger/v3 plugin/storage/badger
plugin/storage/badger/spanstore/rw_internal_test.go:23:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/spanstore/cache_test.go:20:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/spanstore/cache.go:22:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/spanstore/writer.go:24:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/spanstore/reader.go:27:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/factory.go:27:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/samplingstore/storage_test.go:22:	"github.com/dgraph-io/badger/v3"
plugin/storage/badger/samplingstore/storage.go:23:	"github.com/dgraph-io/badger/v3"
```

They may be compatibility issues both in the APIs and with the data. In particular, we will need to verify that data created with the current version can still be used with v4.

### PR DESCRIPTION
**Pull request by hellspawn679**
## Which problem is this PR solving?
- Resolves #5611

## Description of the changes
- updated flag names that was changes from v3 to v4 causing the test's to fail 

## How was this change tested?
- test log in the comments

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
```
Comment (yurishkuro) [inline yarn.lock:2]: We don't need this

File: go.mod [hunk 2]
```diff
@@ -87,6 +87,8 @@ require (
 	gopkg.in/yaml.v3 v3.0.1
 )
 
+require github.com/dgraph-io/badger/v4 v4.2.0 // indirect
```
Comment (yurishkuro) [inline go.mod:90]: This looks odd. Did you start from the renovate bit pr?

File: plugin/storage/badger/factory_test.go [hunk 3]
```diff
@@ -114,7 +114,8 @@ func TestMaintenanceRun(t *testing.T) {
 
 	// This is to for codecov only. Can break without anything else breaking as it does test badger's
 	// internal implementation
-	vlogSize := expvar.Get("badger_v3_vlog_size_bytes").(*expvar.Map).Get(f.tmpDir).(*expvar.Int)
+	// fmt.Println(expvar.Get("badger_size_bytes_vlog"))
```
Comment (yurishkuro) [inline plugin/storage/badger/factory_test.go:117]:
```suggestion
```

File: plugin/storage/badger/factory_test.go [hunk 4]
```diff
@@ -166,28 +167,28 @@ func TestBadgerMetrics(t *testing.T) {
...
 	assert.NotNil(t, f.metrics.badgerMetrics)
-	_, found := f.metrics.badgerMetrics["badger_v3_memtable_gets_total"]
+	_, found := f.metrics.badgerMetrics["badger_get_num_memtable"]
 	assert.True(t, found)
 
 	waiter := func(previousValue int64) int64 {
 		sleeps := 0
 		_, gs := mFactory.Snapshot()
-		for gs["badger_v3_memtable_gets_total"] == previousValue && sleeps < 8 {
+		for gs["badger_get_num_memtable"] == previousValue && sleeps < 8 {
```
Comment (yurishkuro) [inline plugin/storage/badger/factory_test.go:176]: What's the source of these changes? I understand expvar that may be defined by badger lib directly? But this is coming through our metrics api.

**Comment by yurishkuro** [comment]
Please see testing requirements in the issue.

File: yarn.lock (hunk 1 above)
Comment (hellspawn679) [reply yarn.lock:2]: oh sorry that might have been build when I was trying to build jager-ui 
I will remove this

File: plugin/storage/badger/factory_test.go (hunk 4 above)
Comment (hellspawn679) [reply plugin/storage/badger/factory_test.go:176]: the expvar a stored in you go pkg folder so inside that it is specified at `pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.5/y/metrics.go` and for v4 i looked at `pkg/mod/github.com/dgraph-io/badger/v4@v4.2.0/y/metrics.go` for equivalent variable names

Comment (yurishkuro) [reply plugin/storage/badger/factory_test.go:176]: Ah, I see what happens - we have `factory.metricsCopier` function that re-publishes Badger's expvar metrics into our metrics factory

File: yarn.lock (hunk 1 above)
Comment (yurishkuro) [reply yarn.lock]: please also remove bin/golangci-lint

File: go.mod (hunk 2 above)
Comment (yurishkuro) [reply go.mod:90]: https://github.com/jaegertracing/jaeger/pull/5610/files

**Comment by yurishkuro** [comment]
did you do the backwards compatibility test as described in the issue?

**Comment by hellspawn679** [comment]
> did you do the backwards compatibility test as described in the issue?

how can I do that? Is there a make command that I can use to check it?

**Comment by yurishkuro** [comment]
> how can I do that? Is there a make command that I can use to check it?

no, there's not. You'd need:
* run all-in-one with `badger` storage from the `main` branch (or from one of the official released binaries)
  * specify non-ephemeral mode for badger so that it stores data in files on local disk
* run hotrod demo and generate some traces, verify you can load them in the UI. Stop hotrod.
* stop all-in-one and now run it from your branch
* verify you can query & load the same traces in the UI using new version of Badger

**Comment by hellspawn679** [comment]
steps taken :-
1) [ TERMINAL: 1] git checkout main( current main branch) 
2) [ TERMINAL: 1] 
```sh
mkdir tmp && cd tmp && mkdir key && mkdir span
```
image

3)  [ TERMINAL: 1 ] taken form here ([badger](https://www.jaegertracing.io/docs/1.58/cli/#jaeger-all-in-one-badger))
```sh
export SPAN_STORAGE_TYPE=badger
export BADGER_EPHEMERAL=false
export BADGER_CONSISTENCY=true
export BADGER_DIRECTORY_KEY=/home/kali/go/new/jaeger/tmp/key
export BADGER_DIRECTORY_VALUE=/home/kali/go/new/jaeger/tmp/span
cd ..
make run-all-in-one
```
4) [ TERMINAL: 2 ] open a new terminal for hotrod and run it using this command opens at `localhost:8080`
```sh 
go run ./examples/hotrod/main.go all
```
5) click on Rachel's Floral design and click on open trace 
6) this will create new file in `/tmp `dir
image

7) [TERMINAL: 1]we can see span process logs in terminal-1 
image

8) and this is how UI look like
image

9) stop terminal-1 and terminal-2  
10) [TERMINAL: 1]
```sh
git checkout badger
make run-all-in-one
```
11) open` http://localhost:16686/search` select frontend as service and find trace 
image
12) terminal-1 after this 
image

**Comment by yurishkuro** [comment]
that's an awesome log of testing!

**Review by yurishkuro — APPROVED** [review]
Thanks!

**Pull request merged**